bin/inx-chronicle/src/stardust_inx/syncer.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::VecDeque, ops::Range};
+
+use async_trait::async_trait;
+use chronicle::{
+    db::MongoDb,
+    runtime::{Actor, ActorContext, ActorError, ConfigureActor, HandleEvent, Report},
+};
+use inx::{client::InxClient, tonic::Channel};
+
+use super::{InxError, MilestoneStream};
+
+// The Syncer goes backwards in time and tries collect as many milestones as possible.
+pub struct Syncer {
+    gaps: Gaps,
+    db: MongoDb,
+    inx_client: InxClient<Channel>,
+}
+
+impl Syncer {
+    pub fn new(gaps: Vec<Range<u32>>, db: MongoDb, inx_client: InxClient<Channel>) -> Self {
+        Self {
+            gaps: Gaps(gaps.into()),
+            db,
+            inx_client,
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Gaps(VecDeque<Range<u32>>);
+
+impl Iterator for Gaps {
+    type Item = Range<u32>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(range) = self.0.pop_front() {
+                if range.start >= range.end {
+                    continue;
+                } else {
+                    break Some(range);
+                }
+            } else {
+                break None;
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl Actor for Syncer {
+    type State = ();
+    type Error = InxError;
+
+    async fn init(&mut self, _cx: &mut ActorContext<Self>) -> Result<Self::State, Self::Error> {
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl HandleEvent<Report<MilestoneStream>> for Syncer {
+    async fn handle_event(
+        &mut self,
+        cx: &mut ActorContext<Self>,
+        event: Report<MilestoneStream>,
+        _state: &mut Self::State,
+    ) -> Result<(), Self::Error> {
+        // Start syncing the next milestone range
+        cx.delay(SyncNext, None)?;
```

Comment:
What is the reason for the delay here? To slow down syncing in the future?

Reply:
It doesn't actually delay, just sends a message to self.

Reply:
Ah, the good old loop switcheroo.